Research/QuantBook.cs

```diff
@@ -656,5 +766,119 @@ private object GetPropertyValue(object baseData, string fullName)
+
+            // Handle the possible types of conversions
+            if (PyList.IsListType(input))
+            {
+                List<string> symbolsStringList;
+
+                //Check if an entry in the list is a string type, if so then try and convert the whole list
+                if (PyString.IsStringType(input[0]) && input.TryConvert(out symbolsStringList))
+                {
+                    symbolsList = new List<Symbol>();
+                    foreach (var stringSymbol in symbolsStringList)
+                    {
+                        symbol = QuantConnect.Symbol.Create(stringSymbol, SecurityType.Equity, Market.USA);
+                        symbolsList.Add(symbol);
+                    }
+                }
+                //Try converting it to list of symbols, if it fails throw exception
+                else if (!input.TryConvert(out symbolsList))
+                {
+                    throw new Exception($"Cannot convert list {input.Repr()} to symbols");
+                }
+            }
+            else
+            {
+                //Check if its a single string, and try and convert it
+                string symbolString;
+                if (PyString.IsStringType(input) && input.TryConvert(out symbolString))
+                {
+                    symbol = QuantConnect.Symbol.Create(symbolString, SecurityType.Equity, Market.USA);
+                    symbolsList = new List<Symbol> { symbol };
+                }
+                else if (input.TryConvert(out symbol))
+                {
+                    symbolsList = new List<Symbol> { symbol };
+                }
+                else
+                {
+                    throw new Exception($"Cannot convert object {input.Repr()} to symbol");
+                }
+            }
+            return symbolsList;
+        }
+
+        /// <summary>
+        /// Get all fundamental data for given symbols
+        /// </summary>
+        /// <param name="symbols">The symbols to retrieve fundamental data for</param>
+        /// <param name="start">The start date of selected data</param>
+        /// <param name="end">The end date of selected data</param>
+        /// <returns>DataDictionary of Enumerable IBaseData</returns>
+        private Dictionary<Symbol, List<BaseData>> GetAllFundamental(IEnumerable<Symbol> symbols, DateTime? start = null, DateTime? end = null)
+        {
+            //SubscriptionRequest does not except nullable DateTimes, so set a startTime and endTime
+            var startTime = start.HasValue ? (DateTime)start : QuantConnect.Time.BeginningOfTime;
+            var endTime = end.HasValue ? (DateTime)end : QuantConnect.Time.EndOfTime;
+
+            //Collection to store our results
+            var data = new Dictionary<Symbol, List<BaseData>>();
+
+            //Build factory
+            var factory = new FineFundamentalSubscriptionEnumeratorFactory(false);
+            var fileProvider = new DefaultDataProvider();
```

Comment:
We should be using the actual data provider instance here (see `algorithmHandlers.DataProvider`), it might not be the `DefaultDataProvider`

Reply:
Think this guy still has to be replaced by the new `private IDataProvider _dataProvider;`